### ISSUE TITLE
[1.12.2] PackSquash flates out colors in grayscale textures

### ISSUE THREAD

**Issue by realkarmakun** [ComunidadAylas/PackSquash#12]
**Describe the bug**
I tried squashing 1.12.2 resource pack and had texture quality downgraded even with `quantize_image` set to `false`

_Before:_
![изображение](https://user-images.githubusercontent.com/20980281/113480783-f8430580-94b7-11eb-9560-a25a2b0c56bb.png)

_After:_
![изображение](https://user-images.githubusercontent.com/20980281/113480781-f416e800-94b7-11eb-9745-10edee1e2217.png)

**To Reproduce**
1. Squash provided 1.12.2 resource pack with `quantize_image = false`
2. Load it in 1.12.2 Minecraft and see the difference

**Expected behavior**
Squashed resoucre pack textures should look normally (I think)

**Attached files**
My resource packs
[before.zip](https://github.com/ComunidadAylas/PackSquash/files/6252649/before.zip)
[after.zip](https://github.com/ComunidadAylas/PackSquash/files/6252650/after.zip)

**Additional context**
My `config.toml`:
```
resource_pack_directory = "E:\\gstreamer\\RustMeRP"
skip_pack_icon = false
strict_zip_spec_compliance = true
compress_already_compressed_files = false
threads = 4
output_file_path = "E:\\gstreamer\\RustMeRP\\RustMeRP.zip"
["assets/*/sounds/{music,ambience}/?*.{og[ga],mp3,wav,flac}"]
transcode_ogg = false
["**/*.png"]
quantize_image = false
```

**Comment by AlexTMjugador** [comment]
I suspect that a probable cause is that your PNG files contain gamma correction data, which may have been added by your image editing software. Currently, the libraries that PackSquash uses assume that the input PNGs are in the sRGB color space, without gamma correction, but [this will be fixed](https://github.com/ComunidadAylas/PackSquash/blob/09296d0155b3715f56a8bcece30bebe7a700a4ed/src/resource_pack_file/png_file.rs#L184-L193) once v0.3.0 is ready.

If my hypothesis is correct, you may be able to workaround this issue by exporting your PNG files without gamma information. The expected result would be that the texture is identical.

If my hypothesis is not correct, or you're not sure about whether your PNG files contain gamma correction data, please attach a sample PNG file that reproduces this issue so I can analyze it further and see exactly what's going wrong.

**Comment by AlexTMjugador** [comment]
I've been investigating this issue, and I've come to a pretty sound breakthrough of what causes it. I'll share my thoughts and findings in this comment.

<details>
<summary>Investigation process, technical details and breakthrough (advanced stuff)</summary>

> If my hypothesis is correct, you may be able to workaround this issue by exporting your PNG files without gamma information. The expected result would be that the texture is identical.

First things first, this workaround I suggested before may not work because the relevant library that PackSquash uses to do PNG file optimization may still decide that grayscale PNG is the most compact representation for the affected textures. So, if it "fixes" this issue, it'll only be because it tricks PackSquash into not converting the PNGs to grayscale anymore. And, although your PNG files contain gamma correction data, they are just standard sRGB values, which are pretty much assumed everywhere, and shouldn't be the cause of this issue. Still, v0.3.0 should handle non-sRGB color spaces better, by correcting images to sRGB before performing quantization.

I've looked into the deobfuscated Minecraft 1.16.3 client source code for pack handling (both resource and data packs), which resides in classes and packages within `net.minecraft.server.packs`. The resources within a pack are used for a lot of things, but in the case of textures, the class that ends up converting PNG files to textures is `com.mojang.blaze3d.platform.NativeImage`. The relevant code looks like this:

```java
public static NativeImage read(@Nullable NativeImage.Format nativeImageFormat, ByteBuffer imageDataBuffer) throws IOException {
	if (nativeImageFormat != null && !nativeImageFormat.supportedByStb()) {
		throw new UnsupportedOperationException("Don't know how to read format " + nativeImageFormat);
	}
	if (MemoryUtil.memAddress(imageDataBuffer) == 0L) {
		throw new IllegalArgumentException("Invalid buffer");
	}
	try (MemoryStack memoryStack = MemoryStack.stackPush();){
		IntBuffer width = memoryStack.mallocInt(1);
		IntBuffer height = memoryStack.mallocInt(1);
		IntBuffer channels = memoryStack.mallocInt(1);
		ByteBuffer pixelData = STBImage.stbi_load_from_memory(imageDataBuffer, width, height, channels, nativeImageFormat == null ? 0 : nativeImageFormat.components);
		if (pixelData == null) {
			throw new IOException("Could not load image: " + STBImage.stbi_failure_reason());
		}
		NativeImage image = new NativeImage(nativeImageFormat == null ? NativeImage.Format.getStbFormat(channels.get(0)) : nativeImageFormat, width.get(0), height.get(0), true, MemoryUtil.memAddress(pixelData));
		return image;
	}
}
```

As you can see, Minecraft basically delegates all of the heavy-lifting to LWJGL's [`STBImage.stbi_load_from_memory`](https://javadoc.lwjgl.org/org/lwjgl/stb/STBImage.html#stbi_load_from_memory(java.nio.ByteBuffer,java.nio.IntBuffer,java.nio.IntBuffer,java.nio.IntBuffer,int)), which is just a wrapper for the stb_image library, written in C. Therefore, for the purposes of this issue, we're interested in knowing how stb_image deals with converting grayscale (single channel) images to RGBA (four channel). And it turns out that [it does so nicely](https://github.com/nothings/stb/blob/c9064e317699d2e495f36ba4f9ac037e88ee371a/stb_image.h#L1714-L1729), by setting the alpha channel to full opacity, and the rest of channels to the same 8-bit value.

So... what's going wrong, then? Nothing, really. I used your `after.zip` resource pack with Minecraft 1.16.5, and the stone block displayed its texture as you'd expect.

![after.zip working fine in Minecraft 1.16.5](https://user-images.githubusercontent.com/7822554/113510982-bb314e80-955d-11eb-8b7e-c7eafae15182.png)

However, when I opened it with Minecraft 1.12.2, I could reproduce your issue. The stone block looked too bright.

![after.zip not working fine in Minecraft 1.12.2](https://user-images.githubusercontent.com/7822554/113511048-2418c680-955e-11eb-9862-544aae83e31a.png)

This means that Minecraft changed how it reads PNG files between those two versions. Although I didn't examine Minecraft 1.12.2 code because Mojang doesn't provide obfuscation maps for that version, it's pretty safe to assume it used [Java's `ImageIO`](https://docs.oracle.com/javase/8/docs/api/javax/imageio/ImageIO.html#read-java.io.InputStream-) class. ImageIO incorrectly assumes that grayscale PNG files without gamma information have linear gamma, and then corrects that to sRGB, resulting in the washed-out effect we can see ([source](https://stackoverflow.com/questions/31312645/java-imageio-grayscale-png-issue)), which is a pretty identifiable trait of `ImageIO` that I've not seen other libraries doing.

**TL;DR**: old versions of Minecraft used a different library to read PNG images, which makes non-conventional assumptions when computing the color values for grayscale PNG images that do not have color profile data. This results in lighter, "washed-out" colors. This change was made at some point between 1.12.2 and 1.16.3, although probably it was made when 1.13 was released, because in that version Mojang changed a lot of things in the Minecraft engine, and this probably was one of them.
</details>

Now, how can you get your stone texture to show up properly, with the current PackSquash version? The most viable options I see are:

- _Use a newer Minecraft version_. This doesn't happen on 1.16.3 onwards, although you might be able to go back up until 1.13 with the textures still displaying correctly, should you prefer to do so.
- _Introduce a bit of color to your grayscale textures_, ideally in a non-perceptible manner. This way PackSquash won't be able to compress it to grayscale in a lossless manner, because this grayscale format is the root cause of this issue. This is cumbersome and may not work fine with textures that Minecraft expects to be in grayscale, though.
- _Resize your grayscale images to 16x16 pixels_. In my experience, PackSquash doesn't transform grayscale-looking images to actual grayscale format when they are 16x16 pixels or less, because it doesn't yield space savings. YMMV.

In general, any approach that either fixes how Minecraft reads images in grayscale format or does not generate images in grayscale format in the first place will work.

Still, even though PackSquash does things right, I believe it should produce files that do not look, behave or sound wrong on any version of Minecraft that supports the related resource pack feature. To that effect, I'm planning for the next release, v0.3.0, to include a setting like `target_minecraft_version` (this name is not definitive). This setting would enable workarounds for issues like this, forbidding the generation of grayscale PNG files, and would also give enough information to customize PackSquash behavior to better suit the specified version, should I choose to do so. However, its default value should be one that, while being widely compatible with any Minecraft version, assumes that bugs like this, which may impact resource pack size negatively, do not exist.

I still think your feedback is valuable, though. I'd appreciate your confirmation whether things are working for you or not, or whether my findings are correct and explain what are you experiencing. Thanks.

**Comment by realkarmakun** [comment]
Sorry I was busy recently, and wasn't able to check out your answers.

Also amazing research on the issue! A setting for target MC version would be awesome, but of course I will try playing around with grayscale  to make it work with current PackSquash version (maybe in a day or two)

**Comment by realkarmakun** [comment]
Sorry for taking so long. I just tried adding non-gray pixel to the texture, and it does work, you are right about it having to do with grayscale. But gray textures are not the only one that looks different after processing. The grass blocks looks a bit lighter and have less details  (check out Before After in the first message). I will try playing around with textures, but I don't have much time lately (that's why it took me 7 days :D) Thank you for you research though, and all the effort.
![изображение](https://user-images.githubusercontent.com/20980281/114703446-1aa21200-9d47-11eb-9f5b-da0963572706.png)

**Comment by AlexTMjugador** [comment]
You're welcome, and thanks for your time!

I'm also glad my conclusions were right and useful.

Grass still looks different because Minecraft uses a grayscale texture for grass, and then recolors it depending on the biome the grass block is in. Your resource pack contains a "bad" grass grayscale texture, so if you use any of the previously mentioned workarounds to turn it into a "good" texture, you should be fine for now.

**Comment by AlexTMjugador** [comment]
Hello again! I've some news: I've just opened a PR in the OxiPNG repository with a change necessary for PackSquash to address this issue: https://github.com/shssoichiro/oxipng/pull/409, so the seeds needed to roll a fix are planted!

Whether this PR is merged upstream or not, PackSquash will use these changes to OxiPNG to be able to skip grayscale reduction, which should also have the not-so-intended side effect of speeding up the PNG optimization process (because less reductions will be tried).

**Comment by AlexTMjugador** [comment]
v0.3.0-rc.1 (a public pre-release for v0.3.0, the version that fixes this issue) is complete and will be available to download in some minutes, so I'm closing this.
